### ISSUE TITLE
Bug: `no-misused-promises` `getAccessibleSymbolChain` Maximum call stack size exceeded

### ISSUE THREAD

**Issue by gtbuchanan** [typescript-eslint/typescript-eslint#11947]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Issue Description

I was unable to reproduce the problem in the playground, but my repo repository shows it consistently. This same block of code pulled from our larger codebase:

```typescript
import { vi } from 'vitest';

vi.mock('bootstrap-vue-next', async (importActual) => {
  const actual = await importActual<typeof import('bootstrap-vue-next')>();
  return {
    ...actual,
    useModal: vi.fn<typeof actual.useModal>(() => {
      const result = actual.useModal();
      return {
        ...result,
        create: vi.fn<typeof result.create<unknown>>(result.create),
      };
    }),
  };
});
```

Causes the `no-floating-promises` rule to fail with an error:

```text
Oops! Something went wrong! :(

ESLint: 9.39.2

RangeError: Maximum call stack size exceeded
Occurred while linting **redacted**\tseslint-no-misused-promises-max-stack-bug\test\setup.ts:3
Rule: "@typescript-eslint/no-floating-promises"
    at getAccessibleSymbolChain (**redacted**\tseslint-no-misused-promises-max-stack-bug\node_modules\.pnpm\typescript@5.9.3\node_modules\typescript\lib\typescript.js:54805:36)
    at isAnySymbolAccessible (**redacted**\tseslint-no-misused-promises-max-stack-bug\node_modules\.pnpm\typescript@5.9.3\node_modules\typescript\lib\typescript.js:54966:37)
    at isSymbolAccessibleWorker (**redacted**\tseslint-no-misused-promises-max-stack-bug\node_modules\.pnpm\typescript@5.9.3\node_modules\typescript\lib\typescript.js:55022:22)
    at isSymbolAccessible (**redacted**\tseslint-no-misused-promises-max-stack-bug\node_modules\.pnpm\typescript@5.9.3\node_modules\typescript\lib\typescript.js:55011:12)
    at trackExistingEntityName (**redacted**\tseslint-no-misused-promises-max-stack-bug\node_modules\.pnpm\typescript@5.9.3\node_modules\typescript\lib\typescript.js:58148:37)
    at Object.trackExistingEntityName (**redacted**\tseslint-no-misused-promises-max-stack-bug\node_modules\.pnpm\typescript@5.9.3\node_modules\typescript\lib\typescript.js:55396:16)
    at tryVisitTypeQuery (**redacted**\tseslint-no-misused-promises-max-stack-bug\node_modules\.pnpm\typescript@5.9.3\node_modules\typescript\lib\typescript.js:137670:60)
    at visitExistingNodeTreeSymbolsWorker (**redacted**\tseslint-no-misused-promises-max-stack-bug\node_modules\.pnpm\typescript@5.9.3\node_modules\typescript\lib\typescript.js:137878:24)
    at visitExistingNodeTreeSymbols (**redacted**\tseslint-no-misused-promises-max-stack-bug\node_modules\.pnpm\typescript@5.9.3\node_modules\typescript\lib\typescript.js:137626:22)
    at visitNode (**redacted**\tseslint-no-misused-promises-max-stack-bug\node_modules\.pnpm\typescript@5.9.3\node_modules\typescript\lib\typescript.js:95084:19)
```

### Reproduction Repository Link

https://github.com/gtbuchanan/tseslint-no-misused-promises-max-stack-bug

### Repro Steps

1) Clone the repo
2) `pnpm install --frozen-lockfile`
3) `pnpm exec lint`

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `8.52.0` |
| `@typescript-eslint/parser`            | `8.52.0` |
| `TypeScript`                           | `5.9.3` |
| `ESLint`                               | `9.39.2` |
| `node`                                 | `24.12.0` |

**Comment by gtbuchanan** [comment]
I found that removing the type argument from `vi.fn` prevents the error:

```diff
diff --git a/test/setup.ts b/test/setup.ts
index 03ed1b9..75aef80 100644
--- a/test/setup.ts
+++ b/test/setup.ts
@@ -4,7 +4,7 @@ vi.mock('bootstrap-vue-next', async (importActual) => {
   const actual = await importActual<typeof import('bootstrap-vue-next')>();
   return {
     ...actual,
-    useModal: vi.fn<typeof actual.useModal>(() => {
+    useModal: vi.fn(() => {
       const result = actual.useModal();
       return {
         ...result,
```

However, this violates [`vitest/require-mock-type-parameters`](https://github.com/vitest-dev/eslint-plugin-vitest/blob/main/docs/rules/require-mock-type-parameters.md) in our case.

**Comment by StyleShit** [comment]
I think this is a bug in TypeScript's `symbolToNode`. Here's the most minimal reproduction I was able to recreate, unrelated to ts-eslint:

```TS
declare const createObject: () => {
  fn: <T>() => Wrapped<T>;
};

interface Wrapped<P> {
  fn: () => Wrapped<P>;
}

function mockFn<T = any>(fn: (...args: any[]) => T) {
  return fn;
}

mockFn<typeof createObject>(() => {
  const result = createObject();

  return mockFn<typeof result.fn<unknown>>(result.fn);
});

```

This code won't compile and will fail with the same infinite recursion error
Seems like it's due to an infinite types loop that TS can't stop


In your specific case, if you remove the `unknown` from `typeof result.create<unknown>` things seem to work

@bradzacher, I think this PR should be marked as blocked by another pr/upstream, and we could raise an issue in Typescript

We can try to think about a workaround in the meantime, but it feels like an edge that I'm not sure is worth the effort. wdyt?

**Comment by harshit-d3v** [comment]
@StyleShit I think dropping the `<unknown>` is the better workaround between the two options here. It avoids the recursion in the repro without tripping `vitest/require-mock-type-parameters`, whereas removing the type arg from `vi.fn` would.\n\nThat said, I still think it makes sense to handle this on the typescript-eslint side even if the root cause is upstream TypeScript. From the user side this currently shows up as `no-floating-promises` crashing on valid code, and users won’t always have an obvious rewrite available.\n\nI opened #12216 with a narrow guard for the recursive type lookup in `no-floating-promises` plus regression coverage, and CI is green now. If maintainers would rather treat it as upstream-only, happy to adjust, but my take is that avoiding a hard rule crash is still a good win here.

**Comment by JoshuaKGoldberg** [comment]
@harshit-d3v your comment (along with several other of your contributions) appears to violate our [AI Contributions Policy](https://typescript-eslint.io/contributing/ai-policy). You didn't even parse out the `\n\n` in there 😛. I'm hiding it as spam. But feel free to keep posting on-topic comments if you'd like to try again, we'd be happy to have your suggestions be part of the conversation.

---

In more detail: we are an independent open source project with very few resources. All of us on the team are doing this in our spare time. The recent flood of AI-generated content means we have a lot more stuff to review, but the average quality is much lower. The onus cannot be on us to deeply explain all of the things wrong with comments (copypasta issues, relevancy tangents). Please be respectful of our needs as maintainers so that we can help you have you meaningfully participate in the project. 💔

**Comment by JoshuaKGoldberg** [comment]
@StyleShit do you have bandwidth to file that issue on TypeScript?

Unfortunately for us, we do still have to support current/past versions of TS that have the bug. So I think we'll want to work around it 😞.

**Comment by StyleShit** [comment]
> do you have bandwidth

Sure!

Filed https://github.com/microsoft/TypeScript/issues/63441

> I think we'll want to work around it

I'll try to think of a solution (other than a `try... catch`)
<sub>but I'm still not sure whether this very edgy edge case, that has a workaround, might be worth the effort</sub>

**Comment by kirkwaiblinger** [comment]
No stress if `try...catch` is the easiest thing here, either! We've had to do this in other places too for TS bugs. (a recent example being https://github.com/typescript-eslint/typescript-eslint/pull/12150)
